mac identifier: Prefer permanent MAC address

cathay4t opened the pull request:
When searching interface for `InterfaceIdentifier::MacAddress`, we
should try permanent MAC address first.

Also check on interface type matches or not.

Unit test cases included.

ihuguet commented:
Can you explain why permanent MAC address should be prefered? I could arge that if a user sets a MAC address to an interface is because he wants to use it, so it should take precedence.

ihuguet commented on rust/src/lib/ifaces/inter_ifaces.rs:
```diff
@@ -411,6 +421,51 @@ impl Interfaces {
             }
+            // If `permanent_mac_address` got no matches, fallback to
+            // `mac_address`
+            if !has_match {
```
This block is basically duplicating the same operations done above. Maybe we can improve it with a loop?

For example: 
```
for use_permanent in &[true, false] {
    let addr = if use_permanent {
        cur_iface.base_iface().permanent_mac_address
    } else {
        cur_iface.base_iface().mac_address
    }
    if addr
        .as_deref()
         .map(|m| m.to_ascii_uppercase())
         .as_deref()
         == Some(&mac_address)
     { ... }
```

liangwen12year replied:
But the logic seems to be wrong here, we do not know if we should use the `use_permanent` address beforehand until we can not find a match when iterating through the interfaces, then we fallback to mac address.

cathay4t replied:
We need to search all permanent mac address first, then fallback to active mac address.

cathay4t replied:
My bad. Yes. It is doable.

cathay4t committed:
mac identifier: Prefer permanent MAC address

When searching interface for `InterfaceIdentifier::MacAddress`, we
should try permanent MAC address first.

Also check on interface type matches or not.

Unit test cases included.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
New patch uploaded with duplicate code removed.
